File: samples/client/3_1_0_unit_test/java/src/main/java/org/openapijsonschematools/client/schemas/validation/MaxContainsValidator.java
package org.openapijsonschematools.client.schemas.validation;

import org.checkerframework.checker.nullness.qual.Nullable;
import org.openapijsonschematools.client.exceptions.ValidationException;

import java.util.List;

public class MaxContainsValidator implements KeywordValidator {
    public final int maxContains;

    public MaxContainsValidator(int maxContains) {
        this.maxContains = maxContains;
    }

    @Override
    public @Nullable PathToSchemasMap validate(
        ValidationData data
    ) {
        if (!(data.arg() instanceof List)) {
            return null;
        }
        if (data.containsPathToSchemas() == null) {
            return null;
        }
        if (data.containsPathToSchemas().size() > maxContains) {
            throw new ValidationException(
                "Validation failed for maxContains keyword in class="+data.schema().getClass()+
                " at pathToItem="+data.validationMetadata().pathToItem()+". Too many items validated to the contains schema."
            );
        }
        return null;
    }
}

File: samples/client/3_1_0_unit_test/kotlin/src/main/kotlin/org/openapijsonschematools/client/schemas/validation/MaxContainsValidator.kt
package org.openapijsonschematools.client.schemas.validation

import org.openapijsonschematools.client.exceptions.ValidationException

class MaxContainsValidator : KeywordValidator {
    @Throws(ValidationException::class)
    override fun validate(
        data: ValidationData
    ): PathToSchemasMap? {
        val maxContains: Int = data.schema.maxContains ?: return null
        if (data.arg !is List<*>) {
            return null
        }
        val containsPathToSchemas: List<PathToSchemasMap> = data.containsPathToSchemas ?: return null
        if (containsPathToSchemas.size > maxContains) {
            throw ValidationException("""
                |Validation failed for maxContains keyword in class=${data.schema.javaClass}
                | at pathToItem=${data.validationMetadata.pathToItem}. Too many items"
                | validated to the contains schema.
                """.trimMargin()
            )
        }
        return null
    }
}